lib/Sema/CSGen.cpp

```diff
@@ -1115,11 +1115,25 @@ namespace {
       auto baseTy = base->getType();
       auto fnTy = FunctionType::get(inputTv, outputTy);
 
+      // FIXME: This can only happen when diagnostics successfully type-checked
+      // sub-expression of the subscript and mutated AST, but under normal
+      // circumstances subscript should never have InOutExpr as a direct child.
+      // Proper fix for such situation requires preventing diagnostics from
+      // re-writing AST after successful type checking of the sub-expressions.
+      if (isa<InOutExpr>(base)) {
+        // Convert InOutType into LValueType, because base of the
+        // subscript should never be marked as inout, but rather as
+        // @lvalue to denote mutability.
+        if (auto inOutType = baseTy->getAs<InOutType>()) {
+          baseTy = LValueType::get(inOutType->getObjectType());
+        }
+      }
+
```

Comment:
It's simpler to just say:
```
if (auto inoutExpr = dyn_cast<InOutExpr>(base))
  base = inoutExpr->getSubExpr();
```
The subexpr type should be an LValueType when it needs to be, or else something else has gone wrong.